### PR TITLE
Aqt/lansa ssl Add SSL to LANSA templates

### PR DESCRIPTION
**Pull request by robe070**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use a parameter for resource locations with the defaultValue set to resourceGroup().location
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values)
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md

- [ x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

* Add SSL to LANSA templates

### CODE REVIEW

**Comment by robe070** [comment]
The error is due to needing to authorise the purchase of an image. How do we get past this?

**Comment by robe070** [comment]
Can I please have information on how to disable Travis CI, as it was with the previous commits, or how to solve the deployment error that requires authorization of the purchase of an image

File: lansa-vmss-windows-autoscale-sql-database/azuredeploy.json [hunk 1]
```diff
@@ -581,6 +626,11 @@
                                        {
                                           "id": "[concat('/subscriptions/', subscription().subscriptionId,'/resourceGroups/', resourceGroup().name, '/providers/Microsoft.Network/loadBalancers/', variables('loadBalancerName'), '/inboundNatPools/', variables('natPoolName'))]"
                                        }
+                                    ],
+                                    "applicationGatewayBackendAddressPools": [
+                                       {
+                                          "id": "[concat('/subscriptions/', subscription().subscriptionId,'/resourceGroups/', resourceGroup().name, '/providers/Microsoft.Network/applicationGateways/', variables('agName'), '/backendAddressPools/', variables('agBackendAddressPoolName'))]"
```
Comment (bmoore-msft) [inline lansa-vmss-windows-autoscale-sql-database/azuredeploy.json]: remove concat, use resourceId

File: lansa-vmss-windows-autoscale-sql-database/azuredeploy.json [hunk 2]
```diff
@@ -676,6 +726,147 @@
             ]
          }
       },
+      {
+         "type": "Microsoft.Insights/autoscaleSettings",
+         "apiVersion": "2015-04-01",
```
Comment (bmoore-msft) [inline lansa-vmss-windows-autoscale-sql-database/azuredeploy.json]: update apiVersion

File: lansa-vmss-windows-autoscale-sql-database/azuredeploy.json [hunk 3]
```diff
@@ -676,6 +726,147 @@
...
+         "name": "[variables('dbvmssName')]",
+         "location": "[resourceGroup().location]",
```
Comment (bmoore-msft) [inline lansa-vmss-windows-autoscale-sql-database/azuredeploy.json]: Add location parameter - see: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/best-practices.md#parameters

File: lansa-vmss-windows-autoscale-sql-database/azuredeploy.json [hunk 4]
```diff
@@ -676,6 +726,147 @@
...
+         "dependsOn": [
+            "[concat('Microsoft.Compute/virtualMachineScaleSets/', variables('dbvmssName'))]"
+         ],
+         "properties": {
+            "name": "[variables('dbvmssName')]",
+            "targetResourceUri": "[concat('/subscriptions/',subscription().subscriptionId, '/resourceGroups/',  resourceGroup().name, '/providers/Microsoft.Compute/virtualMachineScaleSets/', variables('dbvmssName'))]",
```
Comment (bmoore-msft) [inline lansa-vmss-windows-autoscale-sql-database/azuredeploy.json]: remove concat, use resourceId

File: lansa-vmss-windows-autoscale-sql-database/azuredeploy.json [hunk 5]
```diff
@@ -676,6 +726,147 @@
...
+                     "maximum": "1",
+                     "default": "1"
+                  },
+                  "rules": []
+               }
+            ]
+         }
+      },
+      {
+         "apiVersion": "2015-05-01-preview",
```
Comment (bmoore-msft) [inline lansa-vmss-windows-autoscale-sql-database/azuredeploy.json]: update apiVersion

File: lansa-vmss-windows-autoscale-sql-database/azuredeploy.json [hunk 6]
```diff
@@ -676,6 +726,147 @@
...
+                     "RequestTimeout": 120
+                  }
+               }
+            ],
+            "httpListeners": [
+               {
+                  "name": "appGatewayHttpListener",
+                  "properties": {
+                     "FrontendIPConfiguration": {
+                        "Id": "[concat(variables('agID'), '/frontendIPConfigurations/appGatewayFrontendIP')]"
```
Comment (bmoore-msft) [inline lansa-vmss-windows-autoscale-sql-database/azuredeploy.json]: remove concat, use resourceId

File: lansa-vmss-windows-autoscale-sql-database/azuredeploy.json [hunk 7]
```diff
@@ -676,6 +726,147 @@
...
+                     },
+                     "FrontendPort": {
+                        "Id": "[concat(variables('agID'), '/frontendPorts/appGatewayFrontendPort')]"
```
Comment (bmoore-msft) [inline lansa-vmss-windows-autoscale-sql-database/azuredeploy.json]: remove concat, use resourceId

File: lansa-vmss-windows-autoscale-sql-database/azuredeploy.json [hunk 8]
```diff
@@ -676,6 +726,147 @@
...
+                     },
+                     "Protocol": "Https",
+                     "SslCertificate": {
+                        "Id": "[concat(variables('agID'), '/sslCertificates/appGatewaySslCert')]"
```
Comment (bmoore-msft) [inline lansa-vmss-windows-autoscale-sql-database/azuredeploy.json]: remove concat, use resourceId

File: lansa-vmss-windows-autoscale-sql-database/azuredeploy.json [hunk 9]
```diff
@@ -676,6 +726,147 @@
...
+                  }
+               }
+            ],
+            "requestRoutingRules": [
+               {
+                  "Name": "rule1",
+                  "properties": {
+                     "RuleType": "Basic",
+                     "httpListener": {
+                        "id": "[concat(variables('agID'), '/httpListeners/appGatewayHttpListener')]"
```
Comment (bmoore-msft) [inline lansa-vmss-windows-autoscale-sql-database/azuredeploy.json]: remove concat, use resourceId

File: lansa-vmss-windows-autoscale-sql-database/azuredeploy.json [hunk 10]
```diff
@@ -676,6 +726,147 @@
...
+                     },
+                     "backendAddressPool": {
+                        "id": "[concat(variables('agID'), '/backendAddressPools/', variables('agBackendAddressPoolName'))]"
```
Comment (bmoore-msft) [inline lansa-vmss-windows-autoscale-sql-database/azuredeploy.json]: remove concat, use resourceId

File: lansa-vmss-windows-autoscale-sql-database/azuredeploy.json [hunk 11]
```diff
@@ -676,6 +726,147 @@
...
+                     },
+                     "backendHttpSettings": {
+                        "id": "[concat(variables('agID'), '/backendHttpSettingsCollection/appGatewayBackendHttpSettings')]"
```
Comment (bmoore-msft) [inline lansa-vmss-windows-autoscale-sql-database/azuredeploy.json]: remove concat, use resourceId

**Comment by robe070** [comment]
Thank you for reviewing the changes. Is there a way I may get changes reviewed more quickly in the future?

On your suggested amendments, I am submitting two templates in one pull request and there are changes to the samples in both. Now I'm a complete novice with pull requests, but it seems to me you are ripping out one of my templates. Are you actually trying to tell me that I must submit the changes to the two templates in separate pull requests? That its impossible to submit 2 templates in 1 pull request? (Because I do not understand your comment "...and the existing pattern should be folder into this one (i.e. we don't need separate samples for it) when we're ready to include it"
What existing pattern? What folder? Into what 'one'. And I'm ready now, so when is your ready?

And if the result of your amendments is the deletion of my files, how do I reverse that? To allow it would stop the existing template from working.

On the "remove concat, use resourceId", I could not work out the resource id syntax for the application gateway. It rejected everything I tried. Maybe its because the API Version is incorrect? Which you have asked me to update. Yet when I tried to use the latest there were even more issues. It was extremely frustrating trying to work out the syntax for a trivial change. At least what I have done works. Or has there been an update to the AG since I submitted these changes 3 months ago? Please specify which API Version you are referring to.

**Comment by bmoore-msft** [comment]
@robe070 - the best way to get things to go through quickly is to get a passing CI build - we review those first so we don't spend time reviewing things that won't deploy.

re: one sample per PR, yes, you cannot submit more than one sample in a single PR.  You had 2 and I removed one so we can run test and move forward with merging.  The "second/removed" sample uses a pattern of an existing resource - that pattern can be easily included into a single sample (the one that remains).  So if you want to showcase that scenario, just fold it into this sample.  This sample: https://github.com/Azure/azure-quickstart-templates/tree/master/100-marketplace-sample uses that pattern.  What remains in this PR will deploy as a standalone sample.

For multi-part resources IDs, see: https://docs.microsoft.com/en-us/azure/azure-resource-manager/resource-group-template-functions-resource#resourceid

Here is a sample from this template
resourceId('Microsoft.Network/applicationgateways/frontendIPConfigurations', variables('agName'), 'appGatewayFrontendIP')

**Comment by robe070** [comment]
We have a disconnect.

I'll be concise for brevity and clarity.

1. Allowing your changes to continue will stop one of the templates from working, right now in production. Thats clearly untenable.

2. You are persisting with the idea that I want 2 samples. I do not. I want 2 templates each with 1 sample. Each template is entirely separate from the other. Their only connection is that I authored both of them.

3. I need to completely redo this merge request so that it just contains the 1 template with its 1 sample. And I'm not clear how to back out your changes so that my later update of the MR succeeds. So, how do I do that? Do you need to do it?

4. On the "remove concat, use resourceId", I could not work out the resource id syntax for the application gateway. It rejected everything I tried. Maybe its because the API Version is incorrect? Which you have asked me to update. Yet when I tried to use the latest there were even more issues. It was extremely frustrating trying to work out the syntax for a trivial change. At least what I have done works. Or has there been an update to the AG since I submitted these changes 3 months ago? Please specify which API Version you are referring to. Have you actually used the AG?

5. The message I get from Travis CI is "Travis CI fails due to requiring authorization of a paid image". How do I get around that? Previously an exception was made for my two templates because I am required to use a paid image. That is, Travis CI will ALWAYS fail and I will then be ALWAYS having to wait 3 months to get a response.

**Comment by robe070** [comment]
@bmoore-msft Could I have an update please?

**Comment by bmoore-msft** [comment]
Let's break this down a bit...

First, you can only include changes to one sample (i.e folder) in a PR.  You can remove one of the folders from this PR or close and start new.

If you want to merge the PR, follow the contribution guide and BP for the repo.  If you're having trouble with that, submit the changes se we can see what the problem is.

**Comment by robe070** [comment]
@bmoore-msft 
So I don't feel like you have answered my questions. I will submit another Pull Request, so lets not repeat that again. And I'll not ask the most of the others again, but this one is extremely important for me. Please answer it.

And when I do submit the Pull Request the message I will get from Travis CI is "Travis CI fails due to requiring authorization of a paid image". **How do I get around that?** Previously an exception was made for my two templates because I am required to use a paid image. That is, Travis CI will ALWAYS fail and I will then be ALWAYS having to wait 3 months to get a response.

Clearly I do not want to have to wait again. So please inform me what the process is for this specific issue with Travis CI so that I do not have to wait?

File: lansa-vmss-windows-autoscale-sql-database/azuredeploy.json (hunk 2 above)
Comment (robe070) [reply lansa-vmss-windows-autoscale-sql-database/azuredeploy.json]: 2015-04-01 is the latest apiVersion. When I execute this Powershell
`((Get-AzureRmResourceProvider -ProviderNamespace Microsoft.Insights).ResourceTypes | Where-Object ResourceTypeName -eq autoscalesettings).apiversions`
I get this list
```
2015-04-01
2014-04-01
```

File: lansa-vmss-windows-autoscale-sql-database/azuredeploy.json (hunk 3 above)
Comment (robe070) [reply lansa-vmss-windows-autoscale-sql-database/azuredeploy.json]: The location of the autoscalesettings is in the ResourceGroup. ResourceGroup is already prompted for and I use ResourceGroup to obtain the location.

File: lansa-vmss-windows-autoscale-sql-database/azuredeploy.json (hunk 6 above)
Comment (robe070) [reply lansa-vmss-windows-autoscale-sql-database/azuredeploy.json]: Please provide information, maybe a link to a documentation resource, which describes how to reference the property of a template resource that is not the id of the template resource using the Resource Id. Obtaining the id is trivial. Any other property is opaque. Most of the comments you have made are for such types of references.

**Comment by bmoore-msft** [comment]
Closing abandoned request - please reopen if you can address the feedback.

**Pull request closed without merging**